23.0.0.4 staging (#3067)
OpenLiberty/blogs: posts/2023-05-02-23.0.0.4.adoc

---
layout: post
title: "ARM64 container images in Open Liberty 23.0.0.4"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/mbroz2
author_github: https://github.com/mbroz2
seo-title: ARM64 container images in Open Liberty 23.0.0.4 - OpenLiberty.io

> seo-description: Open Liberty 23.0.0.4 provides better, more consistent logging for the features with which the runtime is running.  Joining the existing AMD64, PPC64LE, and S390X architectures, this release also provides container images for the ARM64 architecture.  This release also includes an important security vulnerability (CVE) fix and many notable bug fixes.

Reviewer: ```suggestion
seo-description: Open Liberty 23.0.0.4 provides better, more consistent logging for the features that are installed on the runtime. Starting with this release, container images for the ARM64 architecture are available, along with the existing AMD64, PPC64LE, and S390X architectures. This release also includes an important security vulnerability (CVE) fix and many notable bug fixes.
```

Reviewer 2: Pulling in the second part only as the first time is specific to 'running' features, not 'installed' features.

> blog_description: Open Liberty 23.0.0.4 provides better, more consistent logging for the features with which the runtime is running.  Joining the existing AMD64, PPC64LE, and S390X architectures, this release also provides container images for the ARM64 architecture.  This release also includes an important security vulnerability (CVE) fix and many notable bug fixes.

Reviewer: ```suggestion
blog_description: Open Liberty 23.0.0.4 provides better, more consistent logging for the features that are installed on the runtime. Starting with this release, container images for the ARM64 architecture are available, along with the existing AMD64, PPC64LE, and S390X architectures. This release also includes an important security vulnerability (CVE) fix and many notable bug fixes.
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= ARM64 container images in Open Liberty 23.0.0.4
Michal Broz <https://github.com/mbroz2>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 23.0.0.4 provides better, more consistent logging for the features with which the runtime is running.  Joining the existing AMD64, PPC64LE, and S390X architectures, this release also provides container images for the ARM64 architecture.  This release also includes an important security vulnerability (CVE) fix and many notable bug fixes.

Reviewer: ```suggestion
Open Liberty 23.0.0.4 provides better, more consistent logging for the features that are installed on the runtime. Starting with this release, container images for the ARM64 architecture are available, along with the existing AMD64, PPC64LE, and S390X architectures. This release also includes an important security vulnerability (CVE) fix and many notable bug fixes.
```

In link:{url-about}[Open Liberty] 23.0.0.4:

* <<arm64, ARM64 container image>>
* <<log, Consistently log features on server start>>
//* <<api, Open Liberty API and SPI documentation>>
* <<CVEs, Security Vulnerability (CVE) Fixes>>
* <<bugs, Notable bug fixes>>

> View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A23004+label%3A%22release+bug%22[23.0.0.4].

Reviewer: Should this sentence clarify that the link goes to GitHub? It looks like we're linking to the bug fixes twice in a row because that's how the list ends too
```suggestion
View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A23004+label%3A%22release+bug%22[23.0.0.4] on GitHub.
```

Reviewer 2: I've been wondering if we just want to delete this one liner when we have the bugs section (and just keep it when we don't include the bugs section)

Check out link:{url-prefix}/blog/?search=release&search!=beta[previous Open Liberty GA release blog posts].

[#run]

== Run your apps using 23.0.0.4

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>23.0.0.4</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[23.0.0.4,)'
}
----

Or if you're using link:{url-prefix}/docs/latest/container-images.html[container images]:

[source]
----
FROM icr.io/appcafe/open-liberty
----

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

[#arm64]
== ARM64 container images

> Starting with the 23.0.0.3 release, Liberty makes available Red Hat Universal Based Images (UBI) for the ARM64 architecture.  Prior to these changes, the provided container images were made available for the AMD64, PPC64LE, and S390X architectures. We recommend using the IBM Container Registry (ICR) as it is not rate limited like Docker Hub.

Reviewer: ```suggestion
Starting with the 23.0.0.4 release, Liberty provides Red Hat Universal Based Images (UBI) for the ARM64 architecture.  Prior to these changes, the provided container images were available for only the AMD64, PPC64LE, and S390X architectures. We recommend using the IBM Container Registry (ICR) to obtain the images because it is not rate-limited like Docker Hub.
```

For more information, reference the link:{url-prefix}/docs/latest/container-images.html[Open Liberty container images] documentation.

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // // 
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/24985
// Contact/Reviewer: brenthdaniel,ReeceNana
// // // // // // // // 
[#log]
== Consistently log features on server start

> In previous versions of Liberty, at server start, the feature manager would log a list of added features in a `CWWKF0012I` audit message and removed features in a `CWWKF0013I` audit message. In the case of a warm start, the feature manager would instead log the entire list of installed features in the `CWWKF0012I` message:

Reviewer: ```suggestion
In previous versions of Liberty, at server start, the feature manager logged a list of added features in a `CWWKF0012I` audit message and removed features in a `CWWKF0013I` audit message. In the case of a warm start, the feature manager instead logged the entire list of installed features in the `CWWKF0012I` message:
```

[source, xml]
----
[AUDIT ] CWWKF0012I: The server installed the following features: [el-3.0, jsp-2.3, servlet-3.1].
----

> This behavior has lead to confusion in the past from users who expect the `CWWKF0012I` to always output the full list of installed features. In the case where the list of features is updated in between server starts, someone could believe that previously installed features are no longer there.

Reviewer: ```suggestion
This behavior has lead to confusion from users who expected the `CWWKF0012I` to always output the full list of enabled features. In the case where the list of features is updated in between server starts, someone might believe that previously enabled features are no longer included.
```

> Starting with 23.0.0.4, Liberty will now consistently provide the list of features that were installed every time the server is started. It will also issue messages to identify features that were added or removed since the last start.

Reviewer: ```suggestion
Starting with 23.0.0.4, Liberty consistently provides the full list of features that are enabled every time the server is started. It also issues messages to identify features that were added or removed since the last start.
```

I might be reading this wrong- it's a bit vague as written. If I've misinterpreted the update I still think we need to be more clear about what has changed.

Reviewer 2: This was a pretty minor change that just updates list of features that are printed when the server starts.  It's largely a serviceability improvement

|http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2023-24998[CVE-2023-24998]
|7.5
|Denial of service
|17.0.0.3 - 23.0.0.3
|Affects the link:{url-prefix}/docs/latest/reference/feature/servlet-3.0[servlet-3.0], link:{url-prefix}/docs/latest/reference/feature/servlet-3.1[servlet-3.1], link:{url-prefix}/docs/latest/reference/feature/servlet-4.0[servlet-4.0], link:{url-prefix}/docs/latest/reference/feature/servlet-5.0[servlet-5.0] and link:{url-prefix}/docs/latest/reference/feature/servlet-6.0[servlet-6.0] features
|===

For a list of past security vulnerability fixes, reference the link:{url-prefix}/docs/latest/security-vulnerabilities.html[Security vulnerability (CVE) list].

[#bugs]
== Notable bugs fixed in this release

We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A23004+label%3A%22release+bug%22[full list of bugs fixed in 23.0.0.4].

* link:https://github.com/OpenLiberty/open-liberty/issues/24585[Insufficient Infinispan cache creation for Liberty httpSessionCache]
+

> We've discovered that the way Liberty creates cache instances with Infinispan can cause multiple applications that use the same virtual host name and context-root, even though they're in entirely separate pods and totally unrelated to each other, to actually share a single Cache instance.

Reviewer: ```suggestion
We've discovered that the way Liberty creates cache instances with Infinispan can cause multiple applications that use the same virtual host name and context-root to share a single Cache instance, even when they're in entirely separate pods and totally unrelated to each other.
```

+

> This causes problems when objects from one app are even just removed by the other app accessing the session. For example:

Reviewer: ```suggestion
This behavior causes problems when objects from one app are removed even just by the other app accessing the session. For example:
```

+
[source, xml]
----
[ERROR ] SESN0304E: An exception occurred when reading in an object of the application data for a session from the cache. The exception is: java.lang.ClassNotFoundException: com.ibm.gs.houston.payment.ach.ibm.forms.ACHCreditForm
[ERROR ] SESN0306E: An exception occurred when invalidating a session in the cache. The exception is: java.lang.RuntimeException: Internal Server Error
[ERROR ] SESN0304E: An exception occurred when reading in an object of the application data for a session from the cache. The exception is: java.lang.ClassNotFoundException: com.ibm.gs.houston.payment.ach.ibm.forms.ACHCreditForm
----
+

> This issue has been resolved by introducing the property `appInCacheName` which, when set to `true`, will include the application name in the corresponding cache.

Reviewer: ```suggestion
We resolved this issue by introducing the `appInCacheName` property, which, when set to `true`, includes the application name in the corresponding cache.
```

* link:https://github.com/OpenLiberty/open-liberty/issues/24915[Server hangs at startup when enabling trace specification: com.ibm.ws.*=all]
+

> When the server is started with `com.ibm.ws.*=all` trace specification, the server hangs, because the application was not loaded yet.

Reviewer: ```suggestion
When the server is started with the `com.ibm.ws.*=all` trace specification, the server hangs because the application was not loaded yet.
```

+
The issue has been resolved and the server no longer hangs even when a very broad trace level is set.

* link:https://github.com/OpenLiberty/open-liberty/issues/24804[Encrypted value for internalClientSecret within oauthProvider does not work]
+

> The `internalClientSecret` and `internalClientId` parameters values within `oauthProvider` configuration of OpenID Connect Server are used to generate tokens for https://openliberty.io/docs/latest/oidc-tools.html#personal[Personal Token Management] and https://openliberty.io/docs/latest/oidc-tools.html#users[Users Token Management] UI pages. However, if the `internalClientSecret` value is encrypted via https://openliberty.io/docs/latest/reference/command/securityUtility-commands.html[`securityUtility` command], the two UI pages no longer work. This is caused by an error that states the `ClientId` and `Secret` are invalid.

Reviewer: ```suggestion
The `internalClientSecret` and `internalClientId` parameter values within the `oauthProvider` element configuration of an OpenID Connect Server are used to generate tokens for https://openliberty.io/docs/latest/oidc-tools.html#personal[Personal Token Management] and https://openliberty.io/docs/latest/oidc-tools.html#users[Users Token Management] UI pages. However, if the `internalClientSecret` value is encrypted by the  https://openliberty.io/docs/latest/reference/command/securityUtility-commands.html[`securityUtility` command], the two UI pages no longer work. This is caused by an error that states the `ClientId` and `Secret` are invalid.
```

+
This issue has been resolved by adding support for encrypting the `internalClientSecret` value.

* link:https://github.com/OpenLiberty/open-liberty/issues/24793[JSP Options to pick up web-ext jsp-attribute values on start up (honor disableTldSearch to improve app start up time)]
+

> The JSP Engine does not pick up the `jsp-attributes` during the initial stage of application start-up when set via `ibm-web-ext.xml` or  `server.xml` files.   However, later phases of the start up did pick up the `jsp-attributes`. 

Reviewer: ```suggestion
The JSP Engine does not pick up the `jsp-attributes` during the initial stage of application startup when set by `ibm-web-ext.xml` or  `server.xml` files.   However, later phases of the start up did pick up the `jsp-attributes`. 
```

+

> This inadvertently slowed application start up because `disableTldSearch` (which defaults to `true`) would not be registered when set to `false` and thus, the JSP Engine would continue it's time consuming search for TLDs in the various application libraries and other locations. 

Reviewer: ```suggestion
This inadvertently slowed application startup because `disableTldSearch` (which defaults to `true`) was not registered when set to `false` and so the JSP Engine continued its time-consuming search for TLDs in the various application libraries and other locations. 
```

+

> The issue has been resolved and the JSP Engine now honors the `jsp-attributes` when set via `ibm-web-ext.xml` or `server.xml` files.

Reviewer: ```suggestion
The issue has been resolved and the JSP Engine now honors the `jsp-attributes` when they are set by `ibm-web-ext.xml` or `server.xml` files.
```

* link:https://github.com/OpenLiberty/open-liberty/issues/24683[Port MYFACES-4594 ]
+

> MYFACES-4594 was opened because by default `org.apache.myfaces.LOG_WEB_CONTEXT_PARAMS` defaults to `dev-only` which is defined here: https://myfaces.apache.org/#/core40

Reviewer: ```suggestion
MYFACES-4594 was opened because `org.apache.myfaces.LOG_WEB_CONTEXT_PARAMS` defaults to `dev-only`, which is defined here: https://myfaces.apache.org/#/core40
```

+
> Indicate if info logging all web config params should be done before initializing the webapp. With 4.0.0, the default changed from auto to dev-only. where logging no longer occurs in the production project stage (new default behavior). True enables logging in the production and development project stages. False disables logging.
+

> If there are multiple applications deployed on a server and not all of them have the https://jakarta.ee/specifications/faces/4.0/apidocs/jakarta/faces/application/projectstage#PROJECT_STAGE_PARAM_NAME[ProjectStage] set to `Development` the context parameters may not work correctly for all the applications deployed in the way that it should.  The first application to start would define the logging that would or wouldn't take place.

Reviewer: ```suggestion
If multiple applications are deployed on a server and not all of them have the https://jakarta.ee/specifications/faces/4.0/apidocs/jakarta/faces/application/projectstage#PROJECT_STAGE_PARAM_NAME[ProjectStage] set to `Development`, the context parameters might not work correctly for all the applications. The first application to start would define the logging that would or wouldn't take place.
```

+

> This issue has been resolved and each application is now able to define if context parameters are logged or not. 

Reviewer: ```suggestion
This issue has been resolved and each application is now able to define whether context parameters are logged or not. 
```

* link:https://github.com/OpenLiberty/open-liberty/issues/24469[Java 11 NoSuchAlgorithmException SHA1PRNG when FIPS enabled TS012071744]
+

> When enabling FIPS feature with the IBM Semeru JDK and OpenID connect client enabled this FFDC exception is emitted:

Reviewer: ```suggestion
When you enable the FIPS feature with the IBM Semeru JDK and OpenID connect client enabled, this FFDC exception is emitted:
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= ARM64 container images in Open Liberty 23.0.0.4
Michal Broz <https://github.com/mbroz2>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

Open Liberty 23.0.0.4 provides better, more consistent logging for the features with which the runtime is running.  Starting with this release, container images for the ARM64 architecture are available, along with the existing AMD64, PPC64LE, and S390X architectures.  This release also includes an important security vulnerability (CVE) fix and many notable bug fixes.

In link:{url-about}[Open Liberty] 23.0.0.4:

* <<arm64, ARM64 container image>>
* <<log, Consistently log features on server start>>
//* <<api, Open Liberty API and SPI documentation>>
* <<CVEs, Security Vulnerability (CVE) Fixes>>
* <<bugs, Notable bug fixes>>

> View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A23004+label%3A%22release+bug%22[23.0.0.4].

Reviewer 2: ```suggestion
```
